Added backend support for cumulative metrics in podlist

PiotrDabkowski opened the pull request:
Added metric download support for all resources and implemented metric download as part of DataSelect pipeline. Currently only pods use this new pipeline.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubernetes/dashboard/1118)
<!-- Reviewable:end -->

bryk commented:
Review status: 0 of 11 files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[a discussion](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtbWdPBFgVwBnn5ZfW:-KOtbWdPBFgVwBnn5ZfX:bw44757) (no related file):*
One thing re "common/".  When you asked whether to move this to common, I meant to move it to "common/metric". Sorry that I didn't mention it. Otherwise we're polluting the commons folder with many disjoint features. 

---

*[src/app/backend/resource/common/aggregation.go, line 57 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtagSXanuVl8dWQpSy:-KOtagSXanuVl8dWQpSz:b-9r1jmb) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/aggregation.go#L57)):*
> ```Go
> 	}
> 	newDataPoints := []DataPoint{}
> 	sort.Sort(Xs) // ensure X data points are sorted
> ```

Why do we have to do this? 

---

*[src/app/backend/resource/common/dataselect.go, line 24 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtbqLNYUXFZklLxMV8:-KOtbqLO4c5qk2M45JAS:bqggk7i) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/dataselect.go#L24)):*
> ```Go
> )
> ```

Hmm... What is this? 

---

*[src/app/backend/resource/common/dataselect.go, line 117 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtbBSdLQbrf3fiIfyV:-KOtbBSeT8E2XkTYx_M8:b-x8z7wu) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/dataselect.go#L117)):*
> ```Go
> 		if err != nil {
> 			// Just log and give up.
> 			log.Printf(`Failed to create heapster selector for resource "%s". Probably you forgot to add pods to the CachedResources.`, metricDataCell.GetResourceSelector().ResourceType)
> ```

"you" who is "you"? 

---

*[src/app/backend/resource/common/dataselectquery.go, line 31 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtcFJIs12Nemjcxe3z:-KOtcFJIs12Nemjcxe4-:b5nfk5p) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/dataselectquery.go#L31)):*
> ```Go
> type MetricQuery struct {
> 	// Metrics to download
> 	MetricNames   []string
> ```

Should this have a separate type, different from string? E.g., MetricName?

---

*[src/app/backend/resource/common/dataselectquery.go, line 34 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtcAWSMwcuXZTn5b8B:-KOtcAWSMwcuXZTn5b8C:by2sgii) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/dataselectquery.go#L34)):*
> ```Go
> 	// Aggregations if provided will slightly slow down execution in case of pods.
> 	// If not provided, sum will be used, together with fastest possible heapster download method.
> 	Aggregations  []string
> ```

Should this have a separate type, different from string? E.g., AggregationType?

---

*[src/app/backend/resource/common/downloader.go, line 83 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtcXGLJ6yadkY2ybXb:-KOtcXGLJ6yadkY2ybXc:b4p7d0e) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/downloader.go#L83)):*
> ```Go
> 	MetricName    string `json:"metricName"`
> 	Label         `json:"label"`
> 	// aggregating function used (if any)
> ```

Should we return it? If client requested 'SUM' then it knows that 'SUM' is returned. 

---

*[src/app/backend/resource/common/downloader.go, line 105 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtd7Fw5CaLsW-r5evP:-KOtd7Fw5CaLsW-r5evQ:b-thv1tr) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/downloader.go#L105)):*
> ```Go
> 	return self
> }
> // Note DO NOT use aggregations other than sum after compressing!
> ```

Why is that? 

---

*[src/app/backend/resource/common/downloader.go, line 118 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtd2U7xywiASd8dApH:-KOtd2U86h9ULx0ywkrv:b4w9gf3) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/downloader.go#L118)):*
> ```Go
> 	return result
> }
> ```

Hmm.. What this function does? 

---

*[src/app/backend/resource/common/downloader.go, line 152 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtcw0Y4-OHX_fGvVyi:-KOtcw0Y4-OHX_fGvVyj:b-hzndkq) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/downloader.go#L152)):*
> ```Go
> }
> 
> func HeapsterSelectorFromNativeResource(resourceType MetricResourceType, namespace string, resourceNames []string) (HeapsterSelector, error) {
> ```

This function name should start with 'New' :) 

---

*[src/app/backend/resource/common/heapsterdatamodel.go, line 25 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtdRpbmgENz3Cn8seV:-KOtdRpcoBE3QolJxocO:bxalbyk) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/heapsterdatamodel.go#L25)):*
> ```Go
> 
> // HeapsterJSONFormat represents format of JSON used by heapster when sending multiple data cells
> type HeapsterJSONAllInOneFormat struct {
> ```

I think you dont need this. We have heapster types imported already. See file named podmetrics.go
```go
	heapster "k8s.io/heapster/metrics/api/v1/types"
```

Can you reuse/commonize code from that place? Basically so that we don't have the same code twice. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1118)*
<!-- Sent from Reviewable.io -->

PiotrDabkowski commented:
Review status: 0 of 14 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtbWdPBFgVwBnn5ZfW:-KPIdieLhBMPpLgn0sI9:bmafxwa) (no related file):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> One thing re "common/".  When you asked whether to move this to common, I meant to move it to "common/metric". Sorry that I didn't mention it. Otherwise we're polluting the commons folder with many disjoint features. 
</details>
@ bryk Done + cleaned up + added tests. PTAL

---

*[src/app/backend/resource/common/aggregation.go, line 57 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtagSXanuVl8dWQpSy:-KPIc6KAHWlkHPAJmD4U:bb1ltr1) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/aggregation.go#L57)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Why do we have to do this? 
</details>
Entries in the map are not sorted, we want to return a list of datapoints sorted by x.

---

*[src/app/backend/resource/common/dataselect.go, line 24 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtbqLNYUXFZklLxMV8:-KPIcLfCfXxeem2q2i-k:b-87j1qu) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/dataselect.go#L24)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Hmm... What is this? 
</details>
Gone.

---

*[src/app/backend/resource/common/dataselect.go, line 117 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtbBSdLQbrf3fiIfyV:-KPIcUdYwArqxL5y-R9s:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/dataselect.go#L117)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> "you" who is "you"? 
</details>
Done.

---

*[src/app/backend/resource/common/dataselectquery.go, line 31 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtcFJIs12Nemjcxe3z:-KPIcWhOAlBlFxaP5Mof:b-guo8jc) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/dataselectquery.go#L31)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Should this have a separate type, different from string? E.g., MetricName?
</details>
I dont think there is a pointl unless we do Internam Metric Name <-> Heapster Metric Name conversion. Currently metric name is provided by the use as string and forwarded directly to heapster download params.

---

*[src/app/backend/resource/common/dataselectquery.go, line 34 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtcAWSMwcuXZTn5b8B:-KPIcfn45Z_4qTaIZcpY:b-qu6n94) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/dataselectquery.go#L34)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Should this have a separate type, different from string? E.g., AggregationType?
</details>
Good point. Done.

---

*[src/app/backend/resource/common/downloader.go, line 83 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtcXGLJ6yadkY2ybXb:-KPIck6f3FHvf0yqg4Sa:b6wkvlf) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/downloader.go#L83)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Should we return it? If client requested 'SUM' then it knows that 'SUM' is returned. 
</details>
Client may requests multiple aggregations at once. For example sum and stdDev in one request. Metrics returned may be in random and this label makes it possible to identify aggregation ised. 

---

*[src/app/backend/resource/common/downloader.go, line 105 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtd7Fw5CaLsW-r5evP:-KPIdBGW5IUpWx6phGel:b86hujz) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/downloader.go#L105)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Why is that? 
</details>
Fixed it so that it uses smart download with any aggregation. 

---

*[src/app/backend/resource/common/downloader.go, line 118 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtd2U7xywiASd8dApH:-KPIdXa6azaHj3uCiC68:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/downloader.go#L118)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Hmm.. What this function does? 
</details>
Done.

---

*[src/app/backend/resource/common/downloader.go, line 152 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtcw0Y4-OHX_fGvVyi:-KPIddo5h3gbxr6OeVRT:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/downloader.go#L152)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> This function name should start with 'New' :) 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1118)*
<!-- Sent from Reviewable.io -->

bryk commented:
Reviewed 7 of 18 files at r2.
Review status: 6 of 13 files reviewed at latest revision, 19 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtbWdPBFgVwBnn5ZfW:-KPMWv0ysQnSllX9qHb3:b-7914pq) (no related file):*
<details><summary><i>Previously, PiotrDabkowski (Piotr Dabkowski) wrote…</i></summary>
> @ bryk Done + cleaned up + added tests. PTAL
</details>
Looks really nice. PTAL

---

*[src/app/backend/resource/common/dataselect.go, line 26 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMIGTaVYQH01Mv7ril:-KPMIGTaVYQH01Mv7rim:b-rq3n5k) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselect.go#L26)):*
> ```Go
> 
> 
> type CachedResources struct {
> ```

Can you comment what this is? 

---

*[src/app/backend/resource/common/dataselect.go, line 43 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMIK8-c2pg0gfzPf1a:-KPMIK8-c2pg0gfzPf1b:b-rq3n5k) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselect.go#L43)):*
> ```Go
> }
> 
> type MetricDataCell interface {
> ```

Can you comment what this is? 

---

*[src/app/backend/resource/common/dataselect.go, line 120 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMJV2E2ofytNiwV4VY:-KPMJV2FxEn3auljHREx:b-n1jfle) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselect.go#L120)):*
> ```Go
> 		if err != nil {
> 			// Just log and give up.
> 			log.Printf(`Failed to create heapster selector for resource "%s". Probably pods were not available in CachedResources.`, metricDataCell.GetResourceSelector().ResourceType)
> ```

Should we throw an error here? Is this a programming error or a thing that may or may not happen at any time? 

---

*[src/app/backend/resource/common/dataselect.go, line 128 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMKXeyETSaXt3s-Lyb:-KPMKXezurbTisPqfi22:b-lz17ay) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselect.go#L128)):*
> ```Go
> 		aggregations = metric.OnlyDefaultAggregation
> 	}
> 	self.CumulativeMetricsPromises = heapsterSelectors.DownloadAndAggregate(client, metricNames, aggregations)
> ```

Should this function just return CumulativeMetricsPromises instead of storing it in `self`? 

---

*[src/app/backend/resource/common/dataselectquery.go, line 31 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtcFJIs12Nemjcxe3z:-KPMKkWdCxIUCfzAUGaY:bv4gjcg) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/dataselectquery.go#L31)):*
<details><summary><i>Previously, PiotrDabkowski (Piotr Dabkowski) wrote…</i></summary>
> I dont think there is a pointl unless we do Internam Metric Name <-> Heapster Metric Name conversion. Currently metric name is provided by the use as string and forwarded directly to heapster download params.
</details>
Ok, makes sense.  

---

*[src/app/backend/resource/common/dataselectquery.go, line 33 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMKmCJUvoIB-Tx_SL5:-KPMKmCJUvoIB-Tx_SL6:bwzeawp) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselectquery.go#L33)):*
> ```Go
> 
> type MetricQuery struct {
> 	// Metrics to download
> ```

Can you comment about what values should go here and where I can find all possible ones? 

---

*[src/app/backend/resource/common/dataselectquery.go, line 35 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPML7w6d895kxpSpX_B:-KPML7w7L6s1e-Fp4Kql:bps5a7a) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselectquery.go#L35)):*
> ```Go
> 	// Metrics to download
> 	MetricNames   []string
> 	// Aggregations if provided will slightly slow down execution in case of pods.
> ```

Can you say in first sentence what happens when I provide aggregations here? And that I can make it empty? 

---

*[src/app/backend/resource/common/dataselectquery.go, line 74 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMLKUxastnbWNm4bUb:-KPMLKUyQzkCiLq5aFf7:b-bhyh8l) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselectquery.go#L74)):*
> ```Go
> 		PaginationQuery: paginationQuery,
> 		SortQuery:       sortQuery,
> 		MetricQuery:      graphQuery,
> ```

You should probably run gofmt over this code. 

---

*[src/app/backend/resource/common/metric/downloader.go, line 28 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMOkqoQGO2SPGJMose:-KPMOkqp6iXGJm8vQ3uW:bdlxk0y) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/metric/downloader.go#L28)):*
> ```Go
> type MetricPromises []MetricPromise
> 
> // GetMetrics returns all metrics from MetricPromises.
> ```

Can you comment on the return values? What does it mean to return nil from here? 

---

*[src/app/backend/resource/common/metric/downloader.go, line 135 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMV1mTP3ETZf8O7edh:-KPMV1mTP3ETZf8O7edi:b7gh7dt) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/metric/downloader.go#L135)):*
> ```Go
> 			if err != nil {
> 				for _, originalMappingIndex := range reverseMapping[selector.Path] {
> 					result[originalMappingIndex].Error <- err
> ```

You're accessing global `result` variable from many threads. This can lead to a race condition, you must not do this. Let's talk offline on how to fix this. 

---

*[src/app/backend/resource/common/metric/downloader.go, line 182 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMVQ-36ms3ejT-yeTe:-KPMVQ-36ms3ejT-yeTf:bvonvft) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/metric/downloader.go#L182)):*
> ```Go
> // removeDuplicates returns a new list of strings with duplicates removed.
> func removeDuplicates(list []string) []string {
> 	dummyMap := map[string]bool{}
> ```

`dummyMap`? Can you give it a better name? Indicating what it does? 

---

*[src/app/backend/resource/common/metric/downloader.go, line 187 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMVdWqWg3nRPQAkmQB:-KPMVdWqWg3nRPQAkmQC:bpl6iwn) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/metric/downloader.go#L187)):*
> ```Go
> 	}
> 	uniqueList := make([]string, len(dummyMap))
> 	i := 0
> ```

Why `i` here? 

---

*[src/app/backend/resource/common/metric/heapsterdatamodel.go, line 25 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtdRpbmgENz3Cn8seV:-KPMWClpbJW6k_knUd_m:b-pdawkd) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/metric/heapsterdatamodel.go#L25)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> I think you dont need this. We have heapster types imported already. See file named podmetrics.go
> ```go
> 	heapster "k8s.io/heapster/metrics/api/v1/types"
> ```
> 
> Can you reuse/commonize code from that place? Basically so that we don't have the same code twice. 
</details>
Any update? 

---

*[src/app/backend/resource/common/metric/metricresourcetype.go, line 7 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMWgwH8mdb4DVA4MuS:-KPMWgwH8mdb4DVA4MuT:by4m4rk) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/metric/metricresourcetype.go#L7)):*
> ```Go
> // List of all resource Types that support metric download.
> const (
> 	ResourceTypeReplicaSet            = "replicasets"
> ```

Can you fold this into `backend/common/resource/types.go`? There's already mapping like this there. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1118)*
<!-- Sent from Reviewable.io -->

bryk commented:
@olekzabl Can you review this?

PiotrDabkowski commented:
Review status: 0 of 85 files reviewed at latest revision, 18 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtbWdPBFgVwBnn5ZfW:-KPNvSVKyZrYsd5olSH5:b-5u17mc) (no related file):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Looks really nice. PTAL
</details>
@bryk  PTAL

---

*[src/app/backend/resource/common/metric/heapsterdatamodel.go, line 25 \[r1\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtdRpbmgENz3Cn8seV:-KPNoObicvA_WIwGD21n:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/49a2da33d685810749d4e2f6a14b1228de701ac9/src/app/backend/resource/common/metric/heapsterdatamodel.go#L25)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Any update? 
</details>
Done.

---

*[src/app/backend/resource/common/dataselect.go, line 26 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMIGTaVYQH01Mv7ril:-KPNvVvOCXw_Dp18th6R:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselect.go#L26)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you comment what this is? 
</details>
Done.

---

*[src/app/backend/resource/common/dataselect.go, line 43 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMIK8-c2pg0gfzPf1a:-KPNqmAkOtMWHfoDebbV:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselect.go#L43)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you comment what this is? 
</details>
Done.

---

*[src/app/backend/resource/common/dataselect.go, line 120 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMJV2E2ofytNiwV4VY:-KPNqnMhw92IJsCRslma:brdxhnh) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselect.go#L120)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Should we throw an error here? Is this a programming error or a thing that may or may not happen at any time? 
</details>
It will happen in case of programming error - that is if the programmer forgets to include pods in CachedResources when trying to calculate HeapsterSelector for some resource which has to be converted to pods (for example deployment).

---

*[src/app/backend/resource/common/dataselect.go, line 128 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMKXeyETSaXt3s-Lyb:-KPNrR0audKq_2RcvvML:b-c8vz4i) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselect.go#L128)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Should this function just return CumulativeMetricsPromises instead of storing it in `self`? 
</details>
I changed name of SelectableData to DataSelector, is that sufficient or should I change it to return promises instead of self? 
This would make interface slightly less clear as we would not be able to do something like this anymore:

processed := SelectableData.Filter().Sort().GetCumulativeMetrics(heapsterClient).Paginate()

---

*[src/app/backend/resource/common/dataselectquery.go, line 33 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMKmCJUvoIB-Tx_SL5:-KPNtUAUTAoMQ64qDquA:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselectquery.go#L33)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you comment about what values should go here and where I can find all possible ones? 
</details>
Done.

---

*[src/app/backend/resource/common/dataselectquery.go, line 35 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPML7w6d895kxpSpX_B:-KPNtvHQHqixm0XBL_eL:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselectquery.go#L35)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you say in first sentence what happens when I provide aggregations here? And that I can make it empty? 
</details>
Done.

---

*[src/app/backend/resource/common/dataselectquery.go, line 74 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMLKUxastnbWNm4bUb:-KPNvYgxL2LWZEFfVz_8:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/dataselectquery.go#L74)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> You should probably run gofmt over this code. 
</details>
Done.

---

*[src/app/backend/resource/common/metric/downloader.go, line 28 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMOkqoQGO2SPGJMose:-KPNv_bZMLUv5Lv7ocum:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/metric/downloader.go#L28)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you comment on the return values? What does it mean to return nil from here? 
</details>
Done.

---

*[src/app/backend/resource/common/metric/downloader.go, line 135 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMV1mTP3ETZf8O7edh:-KPNy9H05uY8KTH4HVaS:b-iggh3s) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/metric/downloader.go#L135)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> You're accessing global `result` variable from many threads. This can lead to a race condition, you must not do this. Let's talk offline on how to fix this. 
</details>
I removed multiple threads, they were not necessary. Now we have one thread but it is slightly more complicated.

---

*[src/app/backend/resource/common/metric/downloader.go, line 182 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMVQ-36ms3ejT-yeTe:-KPNy-eaMp3QmJ1vn_sM:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/metric/downloader.go#L182)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> `dummyMap`? Can you give it a better name? Indicating what it does? 
</details>
Done.

---

*[src/app/backend/resource/common/metric/downloader.go, line 187 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMVdWqWg3nRPQAkmQB:-KPNupXSAavkPUCL9kBJ:b-wspjlp) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/metric/downloader.go#L187)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Why `i` here? 
</details>
I changed it to use append.

---

*[src/app/backend/resource/common/metric/metricresourcetype.go, line 7 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMWgwH8mdb4DVA4MuS:-KPNvIFMy3y3YK-Z0Hsx:b45eotr) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/common/metric/metricresourcetype.go#L7)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you fold this into `backend/common/resource/types.go`? There's already mapping like this there. 
</details>
This mapping is specifically used for resources that support metric download, therefore I put it inside metric folder.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1118)*
<!-- Sent from Reviewable.io -->

bryk commented:
Reviewed 3 of 27 files at r2, 94 of 99 files at r4.
Review status: 82 of 85 files reviewed at latest revision, 10 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtbWdPBFgVwBnn5ZfW:-KPRwVDj6AtD4jt-zngA:b3x9rcd) (no related file):*
<details><summary><i>Previously, PiotrDabkowski (Piotr Dabkowski) wrote…</i></summary>
> @bryk  PTAL
</details>
A few last comments. PTAL 

---

*[src/app/backend/resource/dataselect/dataselect.go, line 30 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRryRqLhATpO2gUb9P:-KPRryRqLhATpO2gUb9Q:bss4962) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/dataselect/dataselect.go#L30)):*
> ```Go
> type CachedResources struct {
> 	Pods []api.Pod
> 	// ...
> ```

Remove? 

---

*[src/app/backend/resource/metric/downloader.go, line 129 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRtx5iwzDqegGrwhTL:-KPRtx5iwzDqegGrwhTM:b5zv38r) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/metric/downloader.go#L129)):*
> ```Go
> 	compressedSelectors, reverseMapping := self.compress()
> 
> 	// collect all the required data (as promises
> ```

Nit. Close parent at the right end.

---

*[src/app/backend/resource/metric/downloader.go, line 130 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRuaGRziLYo1YMEWCS:-KPRuaGRziLYo1YMEWCT:b-aghcpq) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/metric/downloader.go#L130)):*
> ```Go
> 
> 	// collect all the required data (as promises
> 	unassignedResourcePromisesMap := map[int]MetricPromises{}
> ```

Can you comment why this is safe multi threading? 

---

*[src/app/backend/resource/metric/downloader.go, line 130 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRudAaKOaQUeF5kgX_:-KPRudAbwvE-Gsg6InwQ:b-1xs0s) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/metric/downloader.go#L130)):*
> ```Go
> 
> 	// collect all the required data (as promises
> 	unassignedResourcePromisesMap := map[int]MetricPromises{}
> ```

Can you make this a list? 

---

*[src/app/backend/resource/metric/heapsterdatamodel.go, line 24 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRutjMFVL-vEk1qGKa:-KPRutjMFVL-vEk1qGKb:b95cqfv) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/metric/heapsterdatamodel.go#L24)):*
> ```Go
> )
> 
> // HeapsterAllInOneDownloadConfig hold config information specifying whether given native heapster resource type supports list download.
> ```

s/hold/holds 

---

*[src/app/backend/resource/metric/metricresourcetype.go, line 7 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMWgwH8mdb4DVA4MuS:-KPRv6-LVvuqof_VsWFj:b7peqqz) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/metric/metricresourcetype.go#L7)):*
<details><summary><i>Previously, PiotrDabkowski (Piotr Dabkowski) wrote…</i></summary>
> This mapping is specifically used for resources that support metric download, therefore I put it inside metric folder.
</details>
Yes, definitely. But I meant this block of constants. You could move it to types.go and use it in lines 137-152 where same constants are used.

---

*[src/app/backend/resource/pod/podlist.go, line 102 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRvr8iyjIGaIP59mLg:-KPRvr8iyjIGaIP59mLh:b7u439n) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/pod/podlist.go#L102)):*
> ```Go
> 
> 	podCells, cumulativeMetricsPromises := dataselect.GenericDataSelectWithMetrics(toCells(pods), dsQuery, dataselect.NoResourceCache, heapsterClient)
>         pods = fromCells(podCells)
> ```

Formatting? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1118)*
<!-- Sent from Reviewable.io -->

PiotrDabkowski commented:
Review status: 76 of 84 files reviewed at latest revision, 10 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KOtbWdPBFgVwBnn5ZfW:-KPRzr2oowjrmUN7gZHv:bubllvs) (no related file):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> A few last comments. PTAL 
</details>
@bryk Done, PTAL

---

*[src/app/backend/resource/dataselect/dataselect.go, line 30 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRryRqLhATpO2gUb9P:-KPS-HGXCTG5eGzKuoDN:bm52ad7) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/dataselect/dataselect.go#L30)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Remove? 
</details>
Removed and added explanation that CachedResources can be extended to include other resources.

---

*[src/app/backend/resource/metric/downloader.go, line 129 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRtx5iwzDqegGrwhTL:-KPRz0rQ_XVS1dkpyaan:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/metric/downloader.go#L129)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Nit. Close parent at the right end.
</details>
Done.

---

*[src/app/backend/resource/metric/downloader.go, line 130 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRuaGRziLYo1YMEWCS:-KPRz1B0o30EWu0ceADR:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/metric/downloader.go#L130)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you comment why this is safe multi threading? 
</details>
Done.

---

*[src/app/backend/resource/metric/downloader.go, line 130 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRudAaKOaQUeF5kgX_:-KPRz1TR8oXHukbYpqRl:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/metric/downloader.go#L130)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Can you make this a list? 
</details>
Done.

---

*[src/app/backend/resource/metric/heapsterdatamodel.go, line 24 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRutjMFVL-vEk1qGKa:-KPRzB0JON2lQoEAnc4y:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/metric/heapsterdatamodel.go#L24)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> s/hold/holds 
</details>
Done.

---

*[src/app/backend/resource/pod/podlist.go, line 102 \[r4\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPRvr8iyjIGaIP59mLg:-KPRzq14ZQ9BdhXWx35H:b-896fix) ([raw file](https://github.com/kubernetes/dashboard/blob/21bec29f90815671ac0fce49a96c90b734a9ae89/src/app/backend/resource/pod/podlist.go#L102)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Formatting? 
</details>
Done.

---

*[src/app/backend/resource/metric/metricresourcetype.go, line 7 \[r2\]](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-KPMWgwH8mdb4DVA4MuS:-KPS3S2rqESu826TVNM5:b-1mm1gq) ([raw file](https://github.com/kubernetes/dashboard/blob/b6d87b1a3b2d28a6e08568bc0b57ed040b08c800/src/app/backend/resource/metric/metricresourcetype.go#L7)):*
<details><summary><i>Previously, bryk (Piotr Bryk) wrote…</i></summary>
> Yes, definitely. But I meant this block of constants. You could move it to types.go and use it in lines 137-152 where same constants are used.
</details>
Changed to use common.ResourceKind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1118)*
<!-- Sent from Reviewable.io -->

bryk commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I'm very happy with this. Great job! :) 

---

Reviewed 6 of 99 files at r4, 9 of 9 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/dashboard/1118#-:-KPS9tpecbGurrZIdZij:b-m2uet4)*
<!-- Sent from Reviewable.io -->
